File: commands/radio_link_input/src/RadioLinkInput.cpp
#include "radio_link_input/RadioLinkInput.h"
#include <iostream>
#include <cstring>
#include <algorithm>
#include <sys/ioctl.h>     // For ioctl
// #include <asm/termbits.h>  // For termios2
#include <termios.h>       // For tcflush
#include <cmath>           // For std::abs

namespace radio_link_input {

#ifndef TCGETS2
#define TCGETS2     _IOR('T', 0x2A, struct termios2)
#define TCSETS2     _IOW('T', 0x2B, struct termios2)
#endif

#ifndef BOTHER
#define BOTHER 0x00001000
#endif

struct termios2 {
    tcflag_t c_iflag;
    tcflag_t c_oflag;
    tcflag_t c_cflag;
    tcflag_t c_lflag;
    cc_t c_line;
    cc_t c_cc[19];
    speed_t c_ispeed;
    speed_t c_ospeed;
};

RadioLinkInput::RadioLinkInput()
    : Node("radio_link_input"),
      serial_fd_(-1),
      serial_port_("/dev/ttyS0"),  // Default serial port
      baud_rate_(100000),            // SBUS baud rate
      running_(false),
      failsafe_(false),
      frame_lost_(false),
      current_state_(RobotState::STOP) {

    // Declare parameters
    this->declare_parameter<std::string>("serial_port", serial_port_);
    this->declare_parameter<int>("baud_rate", baud_rate_);
    this->declare_parameter<int>("lx_channel", 3);
    this->declare_parameter<int>("ly_channel", 2);
    this->declare_parameter<int>("rx_channel", 0);
    this->declare_parameter<int>("ry_channel", 1);
    this->declare_parameter<int>("mode_0", 9);
    this->declare_parameter<int>("mode_1", 6);
    this->declare_parameter<int>("mode_2", 4);

    // Get parameters
    this->get_parameter("serial_port", serial_port_);
    this->get_parameter("baud_rate", baud_rate_);
    this->get_parameter("lx_channel", channel_mapping_.lx_channel);
    this->get_parameter("ly_channel", channel_mapping_.ly_channel);
    this->get_parameter("rx_channel", channel_mapping_.rx_channel);
    this->get_parameter("ry_channel", channel_mapping_.ry_channel);
    this->get_parameter("mode_0", channel_mapping_.mode_0);
    this->get_parameter("mode_1", channel_mapping_.mode_1);
    this->get_parameter("mode_2", channel_mapping_.mode_2);

    // Create publishers
    // joy_publisher_ = this->create_publisher<sensor_msgs::msg::Joy>("joy", 10);
    inputs_publisher_ = this->create_publisher<control_input_msgs::msg::Inputs>("control_input", 10);

    // Setup serial communication
    if (setupSerial()) {
        running_ = true;
        serial_thread_ = std::thread(&RadioLinkInput::serialReadLoop, this);
        RCLCPP_INFO(this->get_logger(), "RadioLinkInput node started successfully");
    } else {
        RCLCPP_ERROR(this->get_logger(), "Failed to setup serial communication");
    }
}

RadioLinkInput::~RadioLinkInput() {
    running_ = false;
    if (serial_thread_.joinable()) {
        serial_thread_.join();
    }
    if (serial_fd_ >= 0) {
        close(serial_fd_);
    }
}

bool RadioLinkInput::setupSerial() {
    // 打开串口，使用 O_RDONLY | O_NOCTTY | O_NONBLOCK 标志使其以只读、非阻塞模式打开
    serial_fd_ = open(serial_port_.c_str(), O_RDONLY | O_NOCTTY | O_NONBLOCK);
    if (serial_fd_ < 0) {
        RCLCPP_ERROR(this->get_logger(), "Error opening serial port: %s", strerror(errno));
        return false;
    }

    struct termios2 tio { };
    if (0 != ioctl(serial_fd_, TCGETS2, &tio)) {
        RCLCPP_ERROR(this->get_logger(), "Failed to get termios2 attributes");
        close(serial_fd_);
        serial_fd_ = -1;
        return false;
    }

    /**
     * Setting serial port,8E2, non-blocking.100Kbps
     */
    tio.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL
            | IXON);
    tio.c_iflag |= (INPCK | IGNPAR);
    tio.c_oflag &= ~OPOST;
    tio.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
    tio.c_cflag &= ~(CSIZE | CRTSCTS | PARODD | CBAUD);
    /**
     * use BOTHER to specify speed directly in c_[io]speed member
     */
    tio.c_cflag |= (CS8 | CSTOPB | CLOCAL | PARENB | BOTHER | CREAD);
    tio.c_ispeed = 100000;
    tio.c_ospeed = 100000;
    tio.c_cc[VMIN] = 0;
    tio.c_cc[VTIME] = 0;

    if (0 != ioctl(serial_fd_, TCSETS2, &tio)) {
        RCLCPP_ERROR(this->get_logger(), "Failed to set termios2 attributes");
        close(serial_fd_);
        serial_fd_ = -1;
        return false;
    }

    tcflush(serial_fd_, TCIFLUSH);

    RCLCPP_INFO(this->get_logger(), "Serial port %s configured successfully", serial_port_.c_str());
    return true;
}

void RadioLinkInput::serialReadLoop() {
    uint8_t buffer[SBUS_FRAME_SIZE];
    int buffer_index = 0;
    bool in_frame = false;

    while (running_) {
        uint8_t byte;
        ssize_t n = read(serial_fd_, &byte, 1);

        if (n > 0) {
            // Look for start byte (0x0F)
            if (!in_frame && byte == 0x0F) {
                in_frame = true;
                buffer_index = 0;
                buffer[buffer_index++] = byte;
            } else if (in_frame) {
                buffer[buffer_index++] = byte;

                // Check if we have a complete frame
                if (buffer_index >= SBUS_FRAME_SIZE) {
                    in_frame = false;

                    // Process the frame
                    std::lock_guard<std::mutex> lock(data_mutex_);
                    if (parseSBUSFrame(buffer, channels_)) {
                        processChannels();
                        // publishJoyMessage();
                        publishInputsMessage();
                    }
                }
            }
        } else if (n < 0 && errno != EAGAIN) {
            RCLCPP_WARN(this->get_logger(), "Serial read error: %s", strerror(errno));
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }

        // Small delay to prevent busy waiting
        std::this_thread::sleep_for(std::chrono::microseconds(100));
    }
}

bool RadioLinkInput::parseSBUSFrame(const uint8_t* data, uint16_t* channels) {
    // SBUS frame format validation
    if (data[0] != 0x0F || data[24] != 0x00) {
        return false;
    }

    // Extract channel data (11 bits per channel, 16 channels)
    channels[0]  = ((data[1]    | data[2] <<8)                 & 0x07FF);
    channels[1]  = ((data[2]>>3 | data[3] <<5)                 & 0x07FF);
    channels[2]  = ((data[3]>>6 | data[4] <<2 | data[5]<<10)   & 0x07FF);
    channels[3]  = ((data[5]>>1 | data[6] <<7)                 & 0x07FF);
    channels[4]  = ((data[6]>>4 | data[7] <<4)                 & 0x07FF);
    channels[5]  = ((data[7]>>7 | data[8] <<1 | data[9]<<9)    & 0x07FF);
    channels[6]  = ((data[9]>>2 | data[10]<<6)                 & 0x07FF);
    channels[7]  = ((data[10]>>5| data[11]<<3)                 & 0x07FF);
    channels[8]  = ((data[12]   | data[13]<<8)                 & 0x07FF);
    channels[9]  = ((data[13]>>3| data[14]<<5)                 & 0x07FF);
    channels[10] = ((data[14]>>6| data[15]<<2 | data[16]<<10)  & 0x07FF);
    channels[11] = ((data[16]>>1| data[17]<<7)                 & 0x07FF);
    channels[12] = ((data[17]>>4| data[18]<<4)                 & 0x07FF);
    channels[13] = ((data[18]>>7| data[19]<<1 | data[20]<<9)   & 0x07FF);
    channels[14] = ((data[20]>>2| data[21]<<6)                 & 0x07FF);
    channels[15] = ((data[21]>>5| data[22]<<3)                 & 0x07FF);

    // Extract flags
    uint8_t flags = data[23];
    failsafe_ = (flags & 0x08) != 0;
    frame_lost_ = (flags & 0x04) != 0;

    return true;
}

void RadioLinkInput::processChannels() {
    // Determine robot state from mode channel
    uint16_t mode_0 = channels_[channel_mapping_.mode_0];
    uint16_t mode_1 = channels_[channel_mapping_.mode_1];
    uint16_t mode_2 = channels_[channel_mapping_.mode_2];
    current_state_ = determineRobotState(mode_0, mode_1, mode_2);

    // Update state tracking
    last_normal_0_ = normalizeChannelValue(mode_0);
    last_normal_1_ = normalizeChannelValue(mode_1);
    if (current_state_ != RobotState::RUN) last_state_ = current_state_;
}

void RadioLinkInput::publishJoyMessage() {
    auto joy_msg = sensor_msgs::msg::Joy();
    joy_msg.header.stamp = this->now();
    joy_msg.header.frame_id = "radio_link";

    // Convert channels to joystick axes (normalized to -1.0 to 1.0)
    joy_msg.axes.resize(SBUS_NUM_CHANNELS);
    for (int i = 0; i < SBUS_NUM_CHANNELS; ++i) {
        joy_msg.axes[i] = normalizeChannelValue(channels_[i]);
    }

    // Set buttons based on state
    joy_msg.buttons.resize(5); // 5 possible states
    // joy_msg.buttons[static_cast<int>(RobotState::STOP)] = (current_state_ == RobotState::STOP) ? 1 : 0;
    // joy_msg.buttons[static_cast<int>(RobotState::STAND_UP)] = (current_state_ == RobotState::STAND_UP) ? 1 : 0;
    // joy_msg.buttons[static_cast<int>(RobotState::STAND_DOWN)] = (current_state_ == RobotState::STAND_DOWN) ? 1 : 0;
    // joy_msg.buttons[static_cast<int>(RobotState::WALK)] = (current_state_ == RobotState::WALK) ? 1 : 0;
    // joy_msg.buttons[static_cast<int>(RobotState::RUN)] = (current_state_ == RobotState::RUN) ? 1 : 0;

    joy_publisher_->publish(joy_msg);
}

void RadioLinkInput::publishInputsMessage() {
    auto inputs_msg = control_input_msgs::msg::Inputs();
    inputs_msg.command = static_cast<int32_t>(current_state_);

    // Map channels to control inputs
    if (inputs_msg.command == 0) {
        inputs_msg.lx = normalizeChannelValue(channels_[channel_mapping_.lx_channel]);
        inputs_msg.ly = normalizeChannelValue(channels_[channel_mapping_.ly_channel]);
        inputs_msg.rx = normalizeChannelValue(channels_[channel_mapping_.rx_channel]);
        inputs_msg.ry = normalizeChannelValue(channels_[channel_mapping_.ry_channel]);
    }
    inputs_publisher_->publish(inputs_msg);
}

float RadioLinkInput::normalizeChannelValue(uint16_t value) {
    // SBUS range: 172 (min) to 1811 (max), center at 992
    // const float SBUS_MIN = 172.0f;
    // const float SBUS_MAX = 1811.0f;
    // const float SBUS_CENTER = 992.0f;

    // // Clamp value to valid range
    // float clamped = std::max(SBUS_MIN, std::min(SBUS_MAX, static_cast<float>(value)));

    // // Normalize to -1.0 to 1.0
    // if (clamped < SBUS_CENTER) {
    //     return -1.0f + 2.0f * (clamped - SBUS_MIN) / (SBUS_CENTER - SBUS_MIN);
    // } else {
    //     return 2.0f * (clamped - SBUS_CENTER) / (SBUS_MAX - SBUS_CENTER);
    // }
    float clamped = (value - 307.0f) / (1693.0f - 307.0f) * 2.0f - 1.0f;
    if (abs(clamped) < 0.05f) clamped = 0.0f; // Deadzone

    return clamped;
}

RadioLinkInput::RobotState RadioLinkInput::determineRobotState(uint16_t mode_0, uint16_t mode_1, uint16_t mode_2) {
    // Map mode channel value to robot states
    float normal_0 = normalizeChannelValue(mode_0);
    float normal_1 = normalizeChannelValue(mode_1);
    float normal_2 = normalizeChannelValue(mode_2);

    // Calculate changes
    float delta_0 = normal_0 - last_normal_0_;
    float delta_1 = normal_1 - last_normal_1_;
    RobotState update_state = RobotState::RUN;

    // std::cout << "normal_0: " << normal_0 << ", last_normal_0_: " << last_normal_0_ << std::endl;
    // std::cout << "normal_1: " << normal_1 << ", last_normal_1_: " << last_normal_1_ << std::endl;

    auto safe_update = [&](RobotState& state, RobotState target) {
        if (target != last_state_) {
            state = target;
        }
    };

    if (delta_0 > 0.5f) {
        safe_update(update_state, RobotState::STAND);
    }
    else if (delta_0 < -0.5f) {
        safe_update(update_state, RobotState::STOP);
    }
    else if (delta_1 > 1.0f && last_state_ == RobotState::STAND) {
        safe_update(update_state, RobotState::TROT);
    }
    else if (delta_1 < -1.0f && last_state_ == RobotState::TROT) {
        safe_update(update_state, RobotState::STAND);
    }
    else {
    }

    if (static_cast<int>(update_state) != 0) std::cout << "update_state: " << static_cast<int>(update_state) << std::endl;
    return update_state;
}
} // namespace radio_link_input

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);

    auto node = std::make_shared<radio_link_input::RadioLinkInput>();

    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}

File: commands/radio_link_input/scripts/radio_link_input.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.timer import Timer
from control_input_msgs.msg import Inputs
import serial
import threading
import time
import math
import subprocess
import os
import signal
import sys

class RadioLinkInput(Node):
    def __init__(self):
        super().__init__('radio_link_input')

        # Declare parameters (matching C++ defaults)
        self.declare_parameter('serial_port', '/dev/ttyS0')
        self.declare_parameter('baud_rate', 100000)
        self.declare_parameter('lx_channel', 3)
        self.declare_parameter('ly_channel', 2)
        self.declare_parameter('rx_channel', 0)
        self.declare_parameter('ry_channel', 1)
        self.declare_parameter('mode_0', 9)
        self.declare_parameter('mode_1', 6)
        self.declare_parameter('mode_2', 5)
        self.declare_parameter('mode_3', 7)

        # Get parameters
        self.serial_port = self.get_parameter('serial_port').get_parameter_value().string_value
        self.baud_rate = self.get_parameter('baud_rate').get_parameter_value().integer_value
        self.channel_mapping = {
            'lx_channel': self.get_parameter('lx_channel').get_parameter_value().integer_value,
            'ly_channel': self.get_parameter('ly_channel').get_parameter_value().integer_value,
            'rx_channel': self.get_parameter('rx_channel').get_parameter_value().integer_value,
            'ry_channel': self.get_parameter('ry_channel').get_parameter_value().integer_value,
            'mode_0': self.get_parameter('mode_0').get_parameter_value().integer_value,
            'mode_1': self.get_parameter('mode_1').get_parameter_value().integer_value,
            'mode_2': self.get_parameter('mode_2').get_parameter_value().integer_value,
            'mode_3': self.get_parameter('mode_3').get_parameter_value().integer_value,
        }

        # Create publisher
        self.inputs_publisher = self.create_publisher(Inputs, 'control_input', 10)

        # Serial communication
        self.serial_fd = None
        self.running = False
        self.serial_thread = None
        self.data_mutex = threading.Lock()

        # SBUS data
        self.SBUS_FRAME_SIZE = 25
        self.SBUS_NUM_CHANNELS = 16
        self.channels = [0] * self.SBUS_NUM_CHANNELS
        self.failsafe = False
        self.frame_lost = False

        # State machine
        self.RobotState = {
            'RUN': 0,
            'STOP': 1,
            'STAND': 2,
            'TROT': 3,
            'FLY': 4,
            'PARK': 5,
        }
        self.current_state = self.RobotState['STOP']

        # State tracking for determineRobotState
        self.last_normal_0 = 0.0
        self.last_normal_1 = 0.0
        self.last_normal_2 = 0.0
        self.last_normal_3 = 0.0
        self.last_state = self.RobotState['STOP']

        # Controller management (integrated from rc_controller_startup.py)
        self.controller_process = None
        self.controller_running = False
        self.last_ch4_normal = 1.0
        
        # Controller types
        self.ControllerType = {
            'NONE': 0,
            'STOP': 1,
            'MPC': 2,
            'RL_SAR': 3
        }
        self.current_controller_type = self.ControllerType['NONE']

        # ROS2 environment settings
        self.ROS_SETUP = '/opt/ros/jazzy/setup.bash'
        self.WORKSPACE_SETUP = '/home/cat/jazzy_ws/install/setup.bash'
        self.RL_SAR_SETUP = '/home/cat/rl_sar_lite/install/setup.bash'
        self.LAUNCH_FILE = '/home/cat/jazzy_ws/src/quadruped_ros2_control/controllers/ocs2_quadruped_controller/launch/mujoco.launch.py'
        self.RL_SAR_COMMAND = 'ros2 run rl_sar rl_real_go1'

        # Setup serial
        if self.setup_serial():
            self.running = True
            self.serial_thread = threading.Thread(target=self.serial_read_loop)
            self.serial_thread.start()
            self.get_logger().info('RadioLinkInput node started successfully')
        else:
            self.get_logger().error('Failed to setup serial communication')

    def __del__(self):
        self.running = False
        if self.serial_thread and self.serial_thread.is_alive():
            self.serial_thread.join()
        if self.serial_fd:
            self.serial_fd.close()
        # Stop controller on destruction
        self.stop_controller()

    def setup_serial(self):
        try:
            # Use pyserial for serial communication
            self.serial_fd = serial.Serial(
                port=self.serial_port,
                baudrate=self.baud_rate,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_EVEN,
                stopbits=serial.STOPBITS_TWO,
                timeout=0,
                xonxoff=False,
                rtscts=False,
                dsrdtr=False
            )
            self.serial_fd.flushInput()
            self.serial_fd.flushOutput()
            self.get_logger().info(f'Serial port {self.serial_port} configured successfully')
            return True
        except Exception as e:
            self.get_logger().error(f'Error opening serial port: {e}')
            return False

    def serial_read_loop(self):
        buffer = bytearray()
        while self.running:
            try:
                byte = self.serial_fd.read(1)
                if byte:
                    byte = byte[0]
                    # Look for start byte (0x0F)
                    if not buffer and byte == 0x0F:
                        buffer.append(byte)
                    elif buffer:
                        buffer.append(byte)
                        # Check if we have a complete frame
                        if len(buffer) >= self.SBUS_FRAME_SIZE:
                            # Process the frame
                            with self.data_mutex:
                                if self.parse_sbus_frame(buffer):
                                    self.process_channels()
                                    self.publish_inputs_message()
                            buffer.clear()
            except Exception as e:
                self.get_logger().warn(f'Serial read error: {e}')
                time.sleep(0.01)
            time.sleep(0.0001)  # Small delay

    def parse_sbus_frame(self, data):
        # SBUS frame format validation
        if len(data) < self.SBUS_FRAME_SIZE or data[0] != 0x0F or data[24] != 0x00:
            return False

        # Extract channel data (11 bits per channel, 16 channels)
        self.channels[0]  = ((data[1]     | data[2] << 8)                  & 0x07FF)
        self.channels[1]  = ((data[2]>>3  | data[3] << 5)                  & 0x07FF)
        self.channels[2]  = ((data[3]>>6  | data[4] << 2 | data[5] << 10)  & 0x07FF)
        self.channels[3]  = ((data[5]>>1  | data[6] << 7)                  & 0x07FF)
        self.channels[4]  = ((data[6]>>4  | data[7] << 4)                  & 0x07FF)
        self.channels[5]  = ((data[7]>>7  | data[8] << 1 | data[9] << 9)   & 0x07FF)
        self.channels[6]  = ((data[9]>>2  | data[10] << 6)                 & 0x07FF)
        self.channels[7]  = ((data[10]>>5 | data[11] << 3)                 & 0x07FF)
        self.channels[8]  = ((data[12]    | data[13] << 8)                 & 0x07FF)
        self.channels[9]  = ((data[13]>>3 | data[14] << 5)                 & 0x07FF)
        self.channels[10] = ((data[14]>>6 | data[15] << 2 | data[16] << 10) & 0x07FF)
        self.channels[11] = ((data[16]>>1 | data[17] << 7)                 & 0x07FF)
        self.channels[12] = ((data[17]>>4 | data[18] << 4)                 & 0x07FF)
        self.channels[13] = ((data[18]>>7 | data[19] << 1 | data[20] << 9) & 0x07FF)
        self.channels[14] = ((data[20]>>2 | data[21] << 6)                 & 0x07FF)
        self.channels[15] = ((data[21]>>5 | data[22] << 3)                 & 0x07FF)

        # Extract flags
        flags = data[23]
        self.failsafe = (flags & 0x08) != 0
        self.frame_lost = (flags & 0x04) != 0

        return True

    def process_channels(self):
        # Determine robot state from mode channel
        mode_0 = self.channels[self.channel_mapping['mode_0']]
        mode_1 = self.channels[self.channel_mapping['mode_1']]
        mode_2 = self.channels[self.channel_mapping['mode_2']]
        mode_3 = self.channels[self.channel_mapping['mode_3']]
        self.current_state = self.determine_robot_state(mode_0, mode_1, mode_2, mode_3)

        # Update state tracking
        self.last_normal_0 = self.normalize_channel_value(mode_0)
        self.last_normal_1 = self.normalize_channel_value(mode_1)
        self.last_normal_2 = self.normalize_channel_value(mode_2)
        self.last_normal_3 = self.normalize_channel_value(mode_3)
        if self.current_state != self.RobotState['RUN']:
            self.last_state = self.current_state

        # Controller management based on channel 4
        ch4 = self.channels[4]
        ch4_normal = self.normalize_channel_value(ch4)
        delta = ch4_normal - self.last_ch4_normal
        # Determine desired controller type
        desired_controller_type = self.ControllerType['NONE']
        if delta > 0.5 and ch4_normal > 0.5:
            desired_controller_type = self.ControllerType['MPC']
        elif delta < -0.5 and ch4_normal < -0.5:
            desired_controller_type = self.ControllerType['RL_SAR']
        elif abs(ch4_normal) < 0.2:
            # else: stays NONE (middle position = stop)
            desired_controller_type = self.ControllerType['STOP']
        
        # Controller switching logic with safety check
        if desired_controller_type != self.current_controller_type:
            if desired_controller_type == self.ControllerType['NONE']:
                pass
            elif desired_controller_type == self.ControllerType['STOP']:
                # Stop any running controller
                # if self.controller_running:
                self.get_logger().info("Stopping controller (middle position)")
                self.stop_controller()
            elif desired_controller_type == self.ControllerType['MPC']:
                # Start MPC controller (only if not currently running RL_SAR)
                if not self.controller_running or self.current_controller_type != self.ControllerType['RL_SAR']:
                    # if self.controller_running:
                    self.stop_controller()  # Stop current controller first
                    self.get_logger().info("Starting MPC controller")
                    self.start_controller(self.ControllerType['MPC'])
                else:
                    self.get_logger().warn("Cannot switch directly from RL_SAR to MPC. Stop controller first.")
            elif desired_controller_type == self.ControllerType['RL_SAR']:
                # Start RL_SAR controller (only if not currently running MPC)
                if not self.controller_running or self.current_controller_type != self.ControllerType['MPC']:
                    # if self.controller_running:
                    self.stop_controller()  # Stop current controller first
                    self.get_logger().info("Starting RL_SAR controller")
                    self.start_controller(self.ControllerType['RL_SAR'])
                else:
                    self.get_logger().warn("Cannot switch directly from MPC to RL_SAR. Stop controller first.")
        
        self.last_ch4_normal = ch4_normal

    def publish_inputs_message(self):
        inputs_msg = Inputs()
        inputs_msg.command = self.current_state
        if self.current_controller_type == self.ControllerType['MPC']:
            if self.current_state == self.RobotState["FLY"]:
                inputs_msg.command = 5

        # Map channels to control inputs
        if inputs_msg.command == 0:
            inputs_msg.lx = self.normalize_channel_value(self.channels[self.channel_mapping['lx_channel']])
            inputs_msg.ly = self.normalize_channel_value(self.channels[self.channel_mapping['ly_channel']])
            inputs_msg.rx = self.normalize_channel_value(self.channels[self.channel_mapping['rx_channel']])
            inputs_msg.ry = self.normalize_channel_value(self.channels[self.channel_mapping['ry_channel']])
            
            if self.last_state == self.RobotState['STAND']:
                inputs_msg.lx *= 0.5
                inputs_msg.ly *= 0.25
                inputs_msg.rx *= 0.5
                inputs_msg.ry *= 0.5
            elif self.last_state == self.RobotState['TROT']:
                inputs_msg.ly *= 1.0
                if self.current_controller_type == self.ControllerType['RL_SAR']:
                    inputs_msg.ly *= 1.5 if inputs_msg.ly > 0 else 0.5
            elif self.last_state == self.RobotState['FLY']:
                tmp_ly = inputs_msg.ly
                inputs_msg.ly = tmp_ly * 2.0
                if self.current_controller_type == self.ControllerType['RL_SAR']:
                    inputs_msg.ly = tmp_ly * 1 if tmp_ly > 0 else tmp_ly * 0.5
            elif self.last_state == self.RobotState['PARK']:
                tmp_ly = inputs_msg.ly
                if self.current_controller_type == self.ControllerType['RL_SAR']:
                    inputs_msg.ly = tmp_ly * 1 if tmp_ly > 0 else tmp_ly * 0.5
            else:
                pass
        self.inputs_publisher.publish(inputs_msg)

    def normalize_channel_value(self, value):
        clamped = (value - 307.0) / (1693.0 - 307.0) * 2.0 - 1.0
        if abs(clamped) < 0.05:
            clamped = 0.0  # Deadzone
        return clamped

    def determine_robot_state(self, mode_0, mode_1, mode_2, mode_3):
        normal_0 = self.normalize_channel_value(mode_0)
        normal_1 = self.normalize_channel_value(mode_1)
        normal_2 = self.normalize_channel_value(mode_2)
        normal_3 = self.normalize_channel_value(mode_3)

        # Calculate changes
        delta_0 = normal_0 - self.last_normal_0
        delta_1 = normal_1 - self.last_normal_1
        delta_2 = normal_2 - self.last_normal_2
        delta_3 = normal_3 - self.last_normal_3
        update_state = self.RobotState['RUN']

        def safe_update(target):
            nonlocal update_state
            if target != self.last_state:
                update_state = target

        if delta_0 > 0.5:
            safe_update(self.RobotState['STAND'])
        elif delta_0 < -0.5:
            safe_update(self.RobotState['STOP'])
        elif delta_1 > 1.0 and self.last_state != self.RobotState['STOP']:
            safe_update(self.RobotState['TROT'])
        elif delta_1 < -1.0:
            safe_update(self.RobotState['STAND'])
        elif normal_3 < -0.5 and delta_3 < -0.5 and self.last_state != self.RobotState['STOP']:
            safe_update(self.RobotState['FLY'])
        elif normal_3 > 0.5 and delta_3 > 0.5 and self.last_state != self.RobotState['STOP']:
            safe_update(self.RobotState['PARK'])
        elif abs(normal_3) < 0.1 and abs(delta_3) > 0.5:
            safe_update(self.RobotState['STAND'])
        else:
            pass

        if update_state != self.RobotState['RUN']:
            self.get_logger().info(f"State changed to {update_state}")

        return update_state

    def start_controller(self, controller_type):
        if self.controller_running:
            self.get_logger().info("Controller already running")
            return
        
        try:
            if controller_type == self.ControllerType['MPC']:
                # Source ROS2 environment and launch MPC controller
                cmd = f"source {self.ROS_SETUP} && source {self.WORKSPACE_SETUP} && ros2 launch {self.LAUNCH_FILE}"
                self.controller_process = subprocess.Popen(cmd, shell=True, executable='/bin/bash')
                self.current_controller_type = self.ControllerType['MPC']
                self.get_logger().info("MPC Controller started")
            elif controller_type == self.ControllerType['RL_SAR']:
                # Source ROS2 environment and run RL_SAR controller
                cmd = f"source {self.ROS_SETUP} && source {self.WORKSPACE_SETUP} && source {self.RL_SAR_SETUP} && {self.RL_SAR_COMMAND}"
                self.controller_process = subprocess.Popen(cmd, shell=True, executable='/bin/bash')
                self.current_controller_type = self.ControllerType['RL_SAR']
                self.get_logger().info("RL_SAR Controller started")
            else:
                self.get_logger().error(f"Unknown controller type: {controller_type}")
                return
                
            self.controller_running = True
        except Exception as e:
            self.get_logger().error(f"Failed to start controller: {e}")
            self.current_controller_type = self.ControllerType['NONE']

    def stop_controller(self):
        # if not self.controller_running:
        #     self.get_logger().info("Controller not running")
        #     return
        try:
            if self.controller_process:
                self.controller_process.terminate()
                self.controller_process.wait(timeout=10)
            
            # 额外 kill
            result = subprocess.run(["pkill", "-f", "rl_real_go1"], capture_output=True, text=True)
            if result.returncode == 0:
                self.get_logger().info("rl_real_go1 processes killed successfully")
            else:
                self.get_logger().warning(f"Failed to kill rl_real_go1 processes: {result.stderr}")
            result = subprocess.run(["pkill", "-f", "ros2_control_node"], capture_output=True, text=True)
            if result.returncode == 0:
                self.get_logger().info("ros2_control_node processes killed successfully")
            else:
                self.get_logger().warning(f"Failed to kill ros2_control_node processes: {result.stderr}")

            self.controller_running = False
            self.current_controller_type = self.ControllerType['STOP']
            self.get_logger().info("Controller stopped")
            # Kill any remaining rviz processes
            result = subprocess.run(["pkill", "-f", "rviz"], capture_output=True, text=True)
            if result.returncode == 0:
                self.get_logger().info("Rviz processes killed successfully")
            else:
                self.get_logger().warning(f"Failed to kill rviz processes: {result.stderr}")
            # Kill any remaining robot_state_publisher processes
            result = subprocess.run(["pkill", "-f", "robot_state_publisher"], capture_output=True, text=True)
            if result.returncode == 0:
                self.get_logger().info("Robot state publisher processes killed successfully")
            else:
                self.get_logger().warning(f"Failed to kill robot state publisher processes: {result.stderr}")
        except Exception as e:
            self.get_logger().error(f"Failed to stop controller: {e}")

def signal_handler(sig, frame):
    print('Exiting...')
    sys.exit(0)

def main(args=None):
    signal.signal(signal.SIGINT, signal_handler)
    rclpy.init(args=args)
    node = RadioLinkInput()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.running = False  # 停止线程
        if node.serial_thread and node.serial_thread.is_alive():
            node.serial_thread.join(timeout=1.0)  # 等待线程退出
        rclpy.shutdown()

if __name__ == '__main__':
    main()